rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,152 @@
+## Summary
+[summary]: #summary
+
+Remove unneeded version tags from `Bin_prot`-serialized data produced
+by `ppx_version`. Generate version tags on an as-needed basis where
+they're useful. Detect versioning violations in the CI version linter
+in a more useful way.
+
+## Motivation
+[motivation]: #motivation
+
+Versioned data types are built from other versioned types and
+primitive OCaml types. When generating `Bin_prot`-serialized data from
+instances of those type, a version tag is generated not only for the
+containing type, but also for each constituent type. That strategy
+allows determining the version of any piece of the serialized data.
+
+The Jane Street `Async` versioned RPC mechanism used in
+`Mina_networking` already provides versioning of data sent between
+nodes. Therefore, version-tagging of the serialized data is not needed
+for that use case. Removing the version tags will reduce the amount of
+data sent over the network.
+
+We may wish to save serialized data and be able to determine its
+version. For that use case, a single "top" tag for the containing type
+suffices.
+
+Some user-visible data, such as public keys, are formatted using the
+current serialization mechanism, where all contained types have
+version tags. For those few cases, allow use of the existing
+mechanism.
+
+The CI version linter currently compares changes to versioned types
+against the PR base branch. That's more fine-grained than needed, if
+the goal of preventing version changes is to allow interoperability
+with deployed software. Instead, compare the types against the branch 
```

Comment:
What if we make a choice to merge something that introduces an acceptable incompatibility (eg: a rename of a field). Does this mean that every PR will fail this check until that version of the software is released?

Reply:
Yes, although we might be able to get around the problem by canonicalizing types as suggested in #8163.

Reply:
We could change the linter so that it complains only if there's a difference with the release branch AND the base branch. Then you'd require manual merging just once, in the case of a renaming.